casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -91,32 +127,132 @@ object EstimatorHelper {
       ancestorEvents = (ancestors.flatMap(_.getBody.deploys.flatMap(_.deployLog)) ++
         ancestors.flatMap(_.getBody.deploys.flatMap(_.paymentLog)))
         .map(EventConverter.toRspaceEvent)
+        .toSet
 
-      produceEvents = ancestorEvents.collect { case p: Produce => p }
-      consumeEvents = ancestorEvents.collect { case c: Consume => c }
-      commEvents    = ancestorEvents.collect { case c: COMM    => c }
+      allProduceEvents = ancestorEvents.collect { case p: Produce => p }
+      allConsumeEvents = ancestorEvents.collect { case c: Consume => c }
+      allCommEvents    = ancestorEvents.collect { case c: COMM    => c }
+      (volatileCommEvents, nonVolatileCommEvents) = allCommEvents
+        .partition(isVolatile(_, allConsumeEvents, allProduceEvents))
+      producesInVolatileCommEvents = volatileCommEvents.flatMap(_.produces)
+      consumesInVolatileCommEvents = volatileCommEvents.map(_.consume)
+      produceEvents                = allProduceEvents.filterNot(producesInVolatileCommEvents.contains(_))
+      consumeEvents                = allConsumeEvents.filterNot(consumesInVolatileCommEvents.contains(_))
+    } yield (produceEvents, consumeEvents, nonVolatileCommEvents)
+
+  private[this] def extractJoinedChannels(b: BlockEvents): Set[Blake2b256Hash] = {
+    def joinedChannels(consumes: Set[Consume]) =
+      consumes.withFilter(Consume.hasJoins).flatMap(_.channelsHashes)
+    b match {
+      case (_, consumes, comms) =>
+        joinedChannels(consumes) ++ joinedChannels(comms.map(_.consume))
+    }
+  }
 
-      producesInCommEvents = commEvents.flatMap(_.produces)
-      consumeInCommEvents  = commEvents.map(_.consume)
+  private[this] def tuplespaceEventsPerChannel(
+      b: BlockEvents
+  ): Map[Blake2b256Hash, Set[TuplespaceEvent]] =
+    b match {
+      case (produces, consumes, comms) =>
+        val nonPersistentProducesInComms = comms.flatMap(_.produces).filterNot(_.persistent)
+        val nonPersistentConsumesInComms = comms.map(_.consume).filterNot(_.persistent)
+        val freeProduces                 = produces.filterNot(nonPersistentProducesInComms.contains)
+        val freeConsumes                 = consumes.filterNot(nonPersistentConsumesInComms.contains)
+        val produceEvents = freeProduces
+          .map(TuplespaceEvent.from(_))
```

Comment:
formatting?

Reply:
Not sure what you mean, this code has been auto-formatted

Reply:
```
val produceEvents = freeProduces
          .map(TuplespaceEvent.from(_))
```
```
val produceEvents = freeProduces.map(TuplespaceEvent.from(_))
```

Reply:
ok